src/adapter/etl-adapter-google-sheet/src/Flow/ETL/Adapter/GoogleSheet/GoogleSheetLoader.php

```diff
@@ -0,0 +1,136 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Flow\ETL\Adapter\GoogleSheet;
+
+use Flow\ETL\{Adapter\GoogleSheet\RowsNormalizer\EntryNormalizer, FlowContext, Loader, Row\Entry, Rows};
+use Google\Service\Sheets;
+use Google\Service\Sheets\ValueRange;
+
+final class GoogleSheetLoader implements Loader, Loader\Closure
+{
+    private ?int $columnsCount = null;
+
+    private string $dateFormat = 'Y-m-d';
+
+    private string $dateTimeFormat = \DateTimeInterface::ATOM;
+
+    private int $startingRow = 0;
+
+    private string $timeFormat = 'H:i';
+
+    private bool $withHeader = true;
+
+    public function __construct(
+        private readonly Sheets $service,
+        private readonly string $spreadsheetId,
+        private readonly string $sheetName,
+    ) {
+    }
+
+    public function closure(FlowContext $context) : void
+    {
+        $this->startingRow = 0;
+        $this->columnsCount = null;
+    }
+
+    public function load(Rows $rows, FlowContext $context) : void
+    {
+        if (!$rows->count()) {
+            return;
+        }
+
+        if (null === $this->columnsCount) {
+            $this->columnsCount = $rows->first()->entries()->count();
+        }
+
+        $currentRow = $this->calculateCurrentRow($rows);
+        $sheetRange = $this->calculateSheetRange($currentRow);
+
+        $normalizer = new RowsNormalizer(
+            new EntryNormalizer($context->config->caster(), $this->dateTimeFormat, $this->dateFormat, $this->timeFormat)
+        );
+
+        $this->service->spreadsheets_values->update(
+            $this->spreadsheetId,
+            $sheetRange->toString(),
+            new ValueRange(
+                [
+                    'values' => $this->values($rows, $normalizer),
+                    'majorDimension' => 'ROWS',
+                ]
+            ),
+            ['valueInputOption' => 'RAW']
+        );
+
+        $this->startingRow = $currentRow;
+    }
+
+    public function withDateFormat(string $dateFormat) : self
+    {
+        $this->dateFormat = $dateFormat;
+
+        return $this;
+    }
+
+    public function withDateTimeFormat(string $dateTimeFormat) : self
+    {
+        $this->dateTimeFormat = $dateTimeFormat;
+
+        return $this;
+    }
+
+    public function withHeader(bool $withHeader) : self
+    {
+        $this->withHeader = $withHeader;
+
+        return $this;
+    }
+
+    public function withStartingRow(int $startingRow) : self
+    {
+        $this->startingRow = $startingRow;
+
+        return $this;
+    }
+
+    public function withTimeFormat(string $timeFormat) : self
+    {
+        $this->timeFormat = $timeFormat;
+
+        return $this;
+    }
+
+    private function calculateCurrentRow(Rows $rows) : int
+    {
+        return $rows->count() + (0 === $this->startingRow ? ($this->withHeader ? 1 : 0) : $this->startingRow);
+    }
+
+    private function calculateSheetRange(int $currentRow) : SheetRangeCalculator
```

Comment:
hmm the method says `calculte` but it returns `SheetRangeCalculator`, I would rename `SheetRangeCalculator` into `SheetRange`

Reply:
We have `SheetRange` class already, but it's working a bit differently, so that's why I added a new one, as the scope of this PR is getting really wide :)

Reply:
>scope of this PR is getting really wide :) 

Maybe, but I already made that mistake many times in the past that I merged half-baked solutions and at the end I was the one who needed to maintain/improve them over time. So as much as inconvenient and unnecessary it might look, I would rather prefer to merge only complete and usable solutions

Reply:
> We have SheetRange class already, 

Uhm, not great, we should then probably merge them or use another name like `SheetAddress`

Reply:
Renamed.